src/audio/kpb.c

```diff
@@ -897,6 +897,119 @@ static int kpb_reset(struct comp_dev *dev)
+			       uint32_t in_channels)
+{
+	struct comp_data *kpb = comp_get_drvdata(dev);
+	struct audio_stream __sparse_cache *istream = &source->stream;
+	struct audio_stream __sparse_cache *ostream = &sink->stream;
+	uint16_t ch;
+	size_t i;
+
+	AE_SETCBEGIN0(ostream->addr);
+	AE_SETCEND0(ostream->end_addr);
+
+	buffer_stream_invalidate(source, size);
+
+	if (kpb->config.sampling_width == IPC4_DEPTH_16BIT) {
+		const ae_int16 *in_ptr = (const ae_int16 *)istream->r_ptr;
+		ae_int16x4 d16 = AE_ZERO16();
+		const size_t in_offset = in_channels * sizeof(ae_int16);
+		const size_t out_offset = kpb->num_of_sel_mic * sizeof(ae_int16);
+		const size_t samples_per_chan = size / (sizeof(uint16_t) * kpb->num_of_sel_mic);
+		ae_int16 *out_ptr;
+
+		for (ch = 0; ch < kpb->num_of_sel_mic; ch++) {
+			const ae_int16 *input_data = (const ae_int16 *)(in_ptr) + kpb->offsets[ch];
+
+			out_ptr = (ae_int16 *)ostream->w_ptr;
+			out_ptr += ch;
+			for (i = 0; i < samples_per_chan; i++) {
+				AE_L16_XP(d16, input_data, in_offset);
+				AE_S16_0_XC(d16, out_ptr, out_offset);
+			}
+		}
+	} else if (kpb->config.sampling_width == IPC4_DEPTH_32BIT) {
+		const ae_int32 *in_ptr = (const ae_int32 *)istream->r_ptr;
+		ae_int32x2 d32 = AE_ZERO32();
+		const size_t in_offset = in_channels * sizeof(ae_int32);
+		const size_t out_offset = kpb->num_of_sel_mic * sizeof(ae_int32);
+		const size_t samples_per_chan = size / (sizeof(uint32_t) * kpb->num_of_sel_mic);
+		ae_int32 *out_ptr;
+
+		for (ch = 0; ch < kpb->num_of_sel_mic; ch++) {
+			const ae_int32 *input_data = (const ae_int32 *)(in_ptr) + kpb->offsets[ch];
+
+			out_ptr = (ae_int32 *)ostream->w_ptr;
+			out_ptr += ch;
+			for (i = 0; i < samples_per_chan; i++) {
+				AE_L32_XP(d32, input_data, in_offset);
+				AE_S32_L_XC(d32, out_ptr, out_offset);
+			}
+		}
+	}
+}
+#else
+static void kpb_micselect_copy(struct comp_dev *dev, struct comp_buffer __sparse_cache *sink,
+			       struct comp_buffer __sparse_cache *source, size_t size,
+			       uint32_t in_channels)
+{
+	struct comp_data *kpb = comp_get_drvdata(dev);
+	struct audio_stream __sparse_cache *istream = &source->stream;
+	struct audio_stream __sparse_cache *ostream = &sink->stream;
+
+	buffer_stream_invalidate(source, size);
+	size_t out_samples;
+	uint16_t ch;
+
+	if (kpb->config.sampling_width == IPC4_DEPTH_16BIT) {
+		const uint16_t *in_data;
+		uint16_t *out_data;
+		const uint32_t samples_per_chan = size / (sizeof(uint16_t) * kpb->num_of_sel_mic);
+
+		for (ch = 0; ch < kpb->num_of_sel_mic; ch++) {
+			out_samples = 0;
+			in_data = (uint16_t *)istream->r_ptr;
+			out_data = (uint16_t *)ostream->w_ptr;
+
+			for (size_t i = 0; i < samples_per_chan; i++) {
+				if (&out_data[out_samples * kpb->num_of_sel_mic + ch]
+						>= (uint16_t *)ostream->end_addr) {
+					out_data = (uint16_t *)ostream->addr;
+					out_samples = 0;
+				}
+				out_data[out_samples * kpb->num_of_sel_mic + ch] =
+						in_data[i * in_channels + kpb->offsets[ch]];
+				out_samples++;
```

Comment:
you could optimise this a bit by replacing multiplications with additions:
```
for (size_t i = 0; i < samples_per_chan; i += in_channels) {
	...
	out_data[out_samples + ch] = in_data[i + kpb->offsets[ch]];
	out_samples += kpb->num_of_sel_mic;
```

Reply:
and one more with
for (ch = 0; ch < kpb->num_of_sel_mic; ch++) {
			out_samples = 0;
			in_data = (uint16_t *)istream->r_ptr + kpb->offsets[ch];
			out_data = (uint16_t *)ostream->w_ptr + ch;

			for (size_t i = 0; i < samples_per_chan; i += in_channels) {
				...
				out_data[out_samples * kpb->num_of_sel_mic] = in_data[i];

Reply:
@fkwasowi oh, sorry, the loop in my change suggestion wasn't correct, I think it should be
```
for (size_t i = 0; i < samples_per_chan * in_channels; i += in_channels) {
```

Reply:
Done

Reply:
sorry, my second for loop is not right too, it should be 
           for (ch = 0; ch < kpb->num_of_sel_mic; ch++) {
		out_samples = 0;
		in_data = (uint16_t *)istream->r_ptr + kpb->offsets[ch];
		out_data = (uint16_t *)ostream->w_ptr + ch;
		for (size_t i = 0; i < samples_per_chan * in_channels; i += in_channels) {
			...
			out_data[out_samples * kpb->num_of_sel_mic] = in_data[i];